app/backend/approaches/promptmanager.py

```diff
@@ -0,0 +1,76 @@
+import json
+import pathlib
+from dataclasses import dataclass
+
+import prompty
+from openai.types.chat import ChatCompletionMessageParam
+
+
+@dataclass
+class RenderedPrompt:
+    all_messages: list[ChatCompletionMessageParam]
+    system_content: str
+    few_shot_messages: list[ChatCompletionMessageParam]
+    past_messages: list[ChatCompletionMessageParam]
+    new_user_content: str
+
+
+class PromptManager:
+
+    def load_prompt(self, path: str):
+        raise NotImplementedError
+
+    def load_tools(self, path: str):
+        raise NotImplementedError
+
+    def render_prompt(self, prompt, data) -> RenderedPrompt:
+        raise NotImplementedError
+
+
+class PromptyManager(PromptManager):
+
+    PROMPTS_DIRECTORY = pathlib.Path(__file__).parent / "prompts"
+
+    def load_prompt(self, path: str):
+        return prompty.load(self.PROMPTS_DIRECTORY / path)
+
+    def load_tools(self, path: str):
+        return json.loads(open(self.PROMPTS_DIRECTORY / path).read())
+
+    def render_prompt(self, prompt, data) -> RenderedPrompt:
```

Comment:
This was my solution to my conundrum about needing to retrieve the messages back from the rendered prompt using indexes, so that they can then get passed to the token counter.
Now, what I do is that I mark each example with (EXAMPLE), so that I can distinguish examples from actual past messages, and then I can extract them all back out in this function.

So this function extracts:
* system
* (EXAMPLE) pairs
* past messages
* new user message

Reply:
This is also good feedback for prompty as well, thanks for breaking this out